### PR TITLE
[Fix] Unloaded weights will not be initialized when using PretrainedIinit

### LINKED ISSUE #691: Unexpected weight initialization
Thanks for your error report and we appreciate it a lot.

**Checklist**

1. I have searched related issues but cannot get the expected help.
2. I have read the [FAQ documentation](https://mmengine.readthedocs.io/en/latest/faq.html) but cannot get the expected help.
3. The bug has not been fixed in the latest version.

**Describe the bug**

When using `Pretrained` init_cfg to init a model, the outside init_cfg will **disable all submodules' init_weights**.

If the outside model's pretrained weight does not strictly match, e.g. different num_classes, the mismatched submodule's weight will initialize **by pytorch's default init weight logic**, instead of the submodule's init_cfg. 

This lead's to some unacceptable results, e.g. a huge classification loss due to the failed initialization of the classification head when finetuning a model.

**Reproduction**

In mmdet, finetuning rtmdet_l_8xb32-300e_coco

```python
_base_ = './rtmdet_l_8xb32-300e_coco.py'

checkpoint = 'work_dir/rtmdet/rtmdet_l_8xb32-300e_coco/epoch_300.pth'
model = dict(bbox_head=dict(num_classes=1), init_cfg=dict(type='Pretrained', checkpoint=checkpoint))
```

### PR DESCRIPTION
**Pull request by HAOCHENYE**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Resolve #691 .

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmengine/model/base_module.py [hunk 1]
```diff
@@ -118,7 +129,8 @@ def init_weights(self):
                         init_info=f'Initialized by '
                         f'user-defined `init_weights`'
                         f' in {m.__class__.__name__} ')
-
+            if self.init_cfg:
```
Comment (C1rN09) [inline mmengine/model/base_module.py:132]: Why use `self.init_cfg` instead of `pretrained_cfg` as a condition here?

File: mmengine/model/base_module.py [hunk 2]
```diff
@@ -100,14 +100,25 @@ def init_weights(self):
...
                 if isinstance(self.init_cfg, dict):
-                    # prevent the parameters of
-                    # the pre-trained model
-                    # from being overwritten by
-                    # the `init_weights`
-                    if self.init_cfg['type'] == 'Pretrained':
-                        return
+                    init_cfgs = [self.init_cfg]
+
+                # PretrainedInit have higher priority than any other init_cfg.
```
Comment (MeowZheng) [inline mmengine/model/base_module.py:108]: maybe also modify the documentation about this priority

**Review by MeowZheng — DISMISSED** [review]

File: docs/en/advanced_tutorials/initialize.md [hunk 3]
```diff
@@ -107,6 +107,10 @@ toy_net.init_weights()
 If `init_cfg` is a `dict`, `type` means a kind of initializer registered in `WEIGHT_INITIALIZERS`. The `Pretrained` means `PretrainedInit`, which could help us to load the target checkpoint.
 All initializers have the same mapping relationship like `Pretrained` -> `PretrainedInit`, which strips the suffix `Init` of the class name. The `checkpoint` argument of `PretrainedInit` means the path of the checkpoint. It could be a local path or a URL.
 
+```{note}
+:obj:`PretrainedInit` has a higher priority than any other initializer. The loaded pretrained weights will overwrite the previous initialized weights.
+```
+
```
Comment (zhouzaida) [inline docs/en/advanced_tutorials/initialize.md:113]: It seems like this block was not rendered as expected.

File: mmengine/model/base_module.py [hunk 4]
```diff
@@ -100,14 +106,25 @@ def init_weights(self):
...
```
Comment (zhouzaida) [inline mmengine/model/base_module.py:114]:
```suggestion
                # PretrainedInit has higher priority than any other init_cfg.
```

File: mmengine/model/base_module.py [hunk 5]
```diff
@@ -100,14 +106,25 @@ def init_weights(self):
...
+                # Therefore we call initialize PretrainedInit last to overwrite
```
Comment (zhouzaida) [inline mmengine/model/base_module.py:115]:
```suggestion
                # Therefore we initialize `pretrained_cfg` last to overwrite
```

File: mmengine/model/base_module.py [hunk 6]
```diff
@@ -100,14 +106,25 @@ def init_weights(self):
...
+                # the previous initialized weights.
+                # See details in https://github.com/open-mmlab/mmengine/issues/691 # noqa E501
+                other_cfgs = []
+                pretrained_cfg = []
+                for init_cfg in init_cfgs:
+                    assert isinstance(init_cfg, dict)
+                    if init_cfg.get('type') == 'Pretrained':
```
Comment (zhouzaida) [inline mmengine/model/base_module.py:122]: Why does it use `get` method to obtain the type field rather than `init_cfg['type']`?

**Review by zhouzaida — APPROVED** [review]

**Review by zhouzaida — DISMISSED** [review]

**Review by zhouzaida — APPROVED** [review]
